src/Compilers/CSharp/Portable/Lowering/SpillSequenceSpiller.cs

```diff
@@ -710,6 +713,29 @@ public override BoundNode VisitAssignmentOperator(BoundAssignmentOperator node)
             }
 
             return UpdateExpression(builder, node.Update(left, right, node.IsRef, node.Type));
+
+            BoundExpression spillReceiver(BoundFieldAccess field, ref BoundSpillSequenceBuilder leftBuilder)
+            {
+                if (field.FieldSymbol.ContainingType.IsReferenceType)
+                {
+                    // a reference type can always live across await so Spill using leftBuilder
+                    return Spill(leftBuilder, VisitExpression(ref leftBuilder, field.ReceiverOpt));
+                }
+                if (field.ReceiverOpt is BoundArrayAccess arrayAccess)
+                {
+                    // an arrayAccess returns a ref so can only be called after the await, but spill expression and indices
+                    var expression = VisitExpression(ref leftBuilder, arrayAccess.Expression);
+                    expression = Spill(leftBuilder, expression, RefKind.None);
+                    var indices = this.VisitExpressionList(ref leftBuilder, arrayAccess.Indices, forceSpill: true);
+                    return arrayAccess.Update(expression, indices, arrayAccess.Type);
```

Comment:
>return arrayAccess.Update(expression, indices, arrayAccess.Type); [](start = 20, length = 65)

Also, what about the side-effects here. Why they are important in the function above, but not here? Shouldn't the logic be the same?